### PR TITLE
nixos/tailscale: add openFirewall option

### PR DESCRIPTION
**Pull request by Xe**
This is intended to help reduce friction when getting started with Tailscale. This adds the Tailscale port and interface to the allowlist on the firewall so that things Just Work:tm:.

Signed-off-by: Xe Iaso <xe@tailscale.com>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/tailscale.nix [hunk 1]
```diff
@@ -9,6 +9,13 @@ in {
   options.services.tailscale = {
     enable = mkEnableOption "Tailscale client daemon";
 
+    openFirewall = mkOption {
+      type = types.bool;
+      default = true;
```
Comment (pennae) [inline nixos/modules/services/networking/tailscale.nix:14]: the default for these options should be `false` as outlined in #149889 :confused:

File: nixos/modules/services/networking/tailscale.nix [hunk 2]
```diff
@@ -31,6 +38,13 @@ in {
 
   config = mkIf cfg.enable {
     environment.systemPackages = [ cfg.package ]; # for the CLI
+    networking.firewall = mkIf cfg.openFirewall {
+      trustedInterfaces = if cfg.interfaceName != "userspace-networking" then
```
Comment (ckiee) [inline nixos/modules/services/networking/tailscale.nix:42]: Let us have one less `if .. then .. else ..`:
```suggestion
      trustedInterfaces = optional (cfg.interfaceName != "userspace-networking") cfg.interfaceName
```

File: nixos/modules/services/networking/tailscale.nix (hunk 1 above)
Comment (Xe) [reply nixos/modules/services/networking/tailscale.nix:14]:
```suggestion
      default = false;
```

**Comment by Kranzes** [comment]
When is opening the firewall needed? I have tailscale on 5+ NixOS hosts and never had the need to open them.

**Comment by ckiee** [comment]
> When is opening the firewall needed? I have tailscale on 5+ NixOS hosts and never had the need to open them.

@Kranzes Some networks have annoying firewalls that prevent it from punching through properly. If you punch a port through manually on one side you can make it work anyway.

**Comment by Kranzes** [comment]
Won't you also have to enable the port on the router's firewall too? Sorry for my ignorance.

**Comment by ckiee** [comment]
image

**Comment by Xe** [comment]
@Kranzes Tailscale has some aggressive hole punching that makes it Just Work most of the time, however removing variables from the equation certainly makes things easier.

**Comment by Kranzes** [comment]
> @Kranzes Tailscale has some aggressive hole punching that makes it Just Work most of the time, however removing variables from the equation certainly makes things easier.

I saw you using `trustedInterfaces` before with Tailscale, is that also needed?

**Comment by Xe** [comment]
Yes, that allows you to host things that are visible over tailscale without having to open firewall ports.

**Comment by Kranzes** [comment]
Could you maybe create an option (or set it by default?) for that in this PR too?

**Comment by colemickens** [comment]
whoa, trustedInterfaces seems very useful.

**Pull request closed without merging**